chainerx_cc/chainerx/kernel_registry.h

```diff
@@ -10,6 +10,28 @@
 
 namespace chainerx {
 
+namespace internal {
+
+template <typename KeyKernelType>
+const char* GetKeyKernelName();
+template <typename KeyKernelType>
+std::type_index GetKeyKernelTypeIndex();
+
+// Note this macro must be used in `chainerx::internal` namespace.
+#define CHAINERX_REGISTER_KEY_KERNEL(cls, name)    \
```

Comment:
I'd like `INTERNAL`  prefix in this and following macro name since it's only for internal use.

Reply:
This was intended to be an external API. Users may add their custom key kernels using this API. I wanted to make this API be usable in any namespace. However, I needed to change my code since PR before https://github.com/chainer/chainer/pull/7970/commits/9cb16950bb9c7aeb3ca2381708425b1d2ece90c9 didn't work on Travis for older compilers. I added a TODO comment to mention this restriction explicitly. Does this make sense? I have no idea to improve this :(

Reply:
OK seems reasonable